src/HoverOverlay.tsx

```diff
@@ -1,56 +1,29 @@
 import { LoadingSpinner } from '@sourcegraph/react-loading-spinner'
-import { castArray, upperFirst } from 'lodash'
+import { upperFirst } from 'lodash'
 import AlertCircleOutlineIcon from 'mdi-react/AlertCircleOutlineIcon'
 import CloseIcon from 'mdi-react/CloseIcon'
 import InformationOutlineIcon from 'mdi-react/InformationOutlineIcon'
 import * as React from 'react'
-import { MarkedString, MarkupContent, MarkupKind } from 'vscode-languageserver-types'
-import { asError, ErrorLike, isErrorLike } from './errors'
-import { highlightCodeSafe, renderMarkdown, toNativeEvent } from './helpers'
+import { ErrorLike, isErrorLike } from './errors'
+import { toNativeEvent } from './helpers'
 import { HoveredToken } from './token_position'
-import { HoverMerged, LOADING } from './types'
-
-/** The component used to render a link */
-export type LinkComponent = React.ComponentType<{ to: string } & React.HTMLAttributes<HTMLElement>>
+import { HoverAttachment, LOADING } from './types'
 
 /**
- * Uses a placeholder `<button>` or a React Router `<Link>` depending on whether `to` is set.
+ * The component used to render an action.
+ *
+ * @template A The type of an action.
  */
-const ButtonOrLink: React.StatelessComponent<
-    { linkComponent: LinkComponent; to?: string } & React.HTMLAttributes<HTMLElement>
-> = ({ linkComponent, to, children, ...rest }) => {
-    const Link = linkComponent
-    return to ? (
-        <Link to={to} {...rest}>
-            {children}
-        </Link>
-    ) : (
-        <button {...rest}>{children}</button>
-    )
-}
+export type ActionComponent<A> = React.ComponentType<A & React.HTMLAttributes<HTMLElement>>
 
 /**
  * @template C Extra context for the hovered token.
+ * @template D The type of the hover content data.
+ * @template A The type of an action.
  */
-export interface HoverOverlayProps<C = {}> {
+export interface HoverOverlayProps<C extends object, D, A> {
     /** What to show as contents */
-    hoverOrError?: typeof LOADING | HoverMerged | null | ErrorLike // TODO disallow null and undefined
-
-    /**
-     * The URL to jump to on go to definition.
-     * If loaded, is set as the href of the go to definition button.
-     * If LOADING, a loader is displayed on the button.
-     * If null, an info alert is displayed "no definition found".
-     * If an error, an error alert is displayed with the error message.
-     */
-    definitionURLOrError?: typeof LOADING | { jumpURL: string } | null | ErrorLike
-
-    /**
-     * The URL to navigate to to view references.
-     * If loaded, is set as the href of the find-references button.
-     * If undefined or null, the button is disabled.
-     */
-    referencesURL?: string | null
+    hoverOrError?: typeof LOADING | (HoverAttachment & D) | null | ErrorLike // TODO disallow null and undefined
```

Comment:
Why disallow `null` and `undefined`? Those values have semantic meaning to the state of this library.

Reply:
I don't know - that TODO comment saying those should be ignored was there before. I'm going to remove this file anyway, so it's a nil issue (no pun intended).